Crypto Algorithms

budgetpreneur opened the pull request:
feat: Add CryptoModule.

Add crypto module that allows configure SDK to encrypt and decrypt messages.

fix: Improve security of crypto implementation.

Improved security of crypto implementation by adding enhanced AES-CBC cryptor.

fix: Fix Newtonsoft Json vulnerability with MaxDepth and upgrade version for non-PCL.

Fixes Newtonsoft Json vulnerability with MaxDepth and upgrade to version  for non-PCL.

refactor: Add SubscribeKey validation for subscribe feature.

Add SubscribeKey validation for subscribe feature.

MikeDobrzan commented on src/Api/PubnubApi/Security/Crypto/CryptoModule.cs:
```diff
@@ -0,0 +1,132 @@
+{
+    public class CryptoModule
+    {
+        private readonly ICryptor _cryptor;
```
The module should allow for multiple cryptors

budgetpreneur replied:
This module supports multiple cryptors using constructor. First constructor expects new or old cryptor(due to contract ICryptor) and second one LegacyCryptor type.
Current we have only two cryptors.
This module supports any other cryptor as well as long as it implements the contract and pass its type into constructor.

budgetpreneur replied:
refactored to support multiple cryptors.

MikeDobrzan commented on src/Api/PubnubApi/Security/Crypto/Cryptor.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace PubnubApi.Security.Crypto
+{
+    public class Cryptor
```
What is the purpose of this wrapper?

budgetpreneur replied:
this can be removed.  I created this before CryptoModue.

budgetpreneur replied:
deleted this file

MikeDobrzan commented on src/Api/PubnubApi/Security/Crypto/Cryptors/AesCbcCryptor.cs:
```diff
@@ -0,0 +1,152 @@
+    {
+        private const string IDENTIFIER = "ACRH";
+
+        public AesCbcCryptor(string cipherKey, IPubnubLog log, IPubnubUnitTest unitTest): base(cipherKey, true, log, unitTest)
```
Let's not pass UnitTests in constructors - we can invert it by exposing event delegates

budgetpreneur replied:
the only purpose is pass known IV for encryption, so that we can verify the expected output. Any other option?

budgetpreneur replied:
removed UnitTests from constructors.

MikeDobrzan commented on src/Api/PubnubApi/Security/Crypto/Cryptors/CryptorBase.cs:
```diff
@@ -0,0 +1,106 @@
+        private readonly bool _useDynamicRandomIV;
+        private readonly IPubnubLog _log;
+        private readonly IPubnubUnitTest _unitTest;
+        protected CryptorBase(string cipherKey, bool useDynamicRandomIV, IPubnubLog log, IPubnubUnitTest unitTest)
```
Let's not pass UnitTests in constructors - we can invert it by exposing event delegates

budgetpreneur replied:
the only purpose is pass known IV for encryption, so that we can verify the expected output. Any other option?

budgetpreneur replied:
removed UnitTests from constructors.

MikeDobrzan commented on src/Api/PubnubApi/Security/Crypto/CryptoModule.cs:
```diff
@@ -0,0 +1,132 @@
+            _cryptor = cryptor;
+            _fallbackCryptor = fallbackCryptor;
+        }
+        public static LegacyCryptor CreateLegacyCryptor(string cipherKey, bool useDynamicIV, IPubnubLog log, IPubnubUnitTest unitTest)
```
Let's not pass UnitTests in constructors - we can invert it by exposing event delegates

budgetpreneur replied:
the only purpose is pass known IV for encryption, so that we can verify the expected output. Any other option?

budgetpreneur replied:
removed UnitTests from constructors.

MikeDobrzan commented on src/Api/PubnubApi/Security/Crypto/CryptoModule.cs:
```diff
@@ -0,0 +1,132 @@
+        {
+            return new LegacyCryptor(cipherKey, true);
+        }
+        public static AesCbcCryptor CreateAesCbcCryptor(string cipherKey, IPubnubLog log, IPubnubUnitTest unitTest)
```
Let's not pass UnitTests in constructors - we can invert it by exposing event delegates

budgetpreneur replied:
the only purpose is pass known IV for encryption, so that we can verify the expected output. Any other option?

budgetpreneur replied:
removed UnitTests from constructors.

MikeDobrzan reviewed (changes requested):
Module should support multiple cryptors. Also, we should not require to pass a unit test class - especially in publicly facing constructor and factory functions.

budgetpreneur commented:
> Module should support multiple cryptors. Also, we should not require to pass a unit test class - especially in publicly facing constructor and factory functions.

Added support with multiple cryptors. Also removed unit test class as param in the public facing constructors.

MikeDobrzan reviewed (approved):
LGTM

MikeDobrzan reviewed (approved):
LGTM

budgetpreneur commented:
@pubnub-release-bot release as v6.19.0.0

parfeon commented:
@pubnub-release-bot release as v6.19.0.0

parfeon commented:
@pubnub-release-bot release as v6.19.0.0
